docs/architecture/ADR-004-interest-and-reserves.md

```diff
@@ -0,0 +1,257 @@
+# ADR 004: Borrow interest implementation and reserves
+
+## Changelog
+
+- October 14, 2021: Initial Draft (@toteki)
+
+## Status
+
+Proposed
+
+## Context
+
+Borrow positions on Umee accrue interest over time.
+When interest accrues, the sum of all assets owed by all users increases for each borrowed token denomination. The amount of that increase serves to benefit lenders (by increasing the token:uToken exchange rate), and also to increase the amount of base assets the Umee system holds in reserve.
+
+The mechanism by which interest is calculated, and then split between incentivizing lenders as per [ADR-001](./ADR-001-interest-stream.md) and reserves as defined in this ADR, will follow.
+
+
+## Decision
+
+Interest on all open borrow positions will be applied every X blocks during `EndBlock`, where X is a governance parameter called `BorrowInterestEpoch`.
+
+Reserves are implemented as a per-token value attached to the `x/leverage` module account.  The module account also holds the lending pool of base assets. As an example, if the module account contains 2000 `uatom` and its `uatom ReserveAmount` is 500, then only 1500 of its `uatom` are available for borrowing or withdrawal.
+
+A governance parameter `ReserveFactor` must be kept which specifies the portion of borrow interest that will be funnelled into reserves. This parameter will either be decided per token, or as a single value for all asset types.
+
+> TODO: Decide on the above.
+
+Both the reserved amount of a given token, and the token:uToken exchange rate, increase when interest is _accrued_, rather then when it is _repaid_.
+
+This timing of reserve increases matches the behavior of the [Compound cToken smart contract](https://github.com/compound-finance/compound-protocol/blob/master/contracts/CToken.sol) we are using as a reference - see around line 410 in function accrueInterest.
+
+## Detailed Design
+
+As noted in [ADR-003](./ADR-003-borrow-assets.md), open borrow positions are stored in the`x/leverage` module with the keys `borrowPrefix | lengthPrefixed(borrowerAddress) | tokenDenom` and values of type `sdk.Int`.
+
+When accruing interest, the borrowed amount (`sdk.Int`) must be increased for each open borrow position. The increase should be calculated as follows...
+
+### Borrow Interest Epoch
+
+> accrued interest = borrowed amount × (interest rate × time elapsed)
+
+A governance parameter `BorrowInterestEpoch` with type `int64` will determine the number blocks to wait between interest calculations. Whenever the number of blocks created since the last interest accrual reaches the desired value, interest accrues on all borrow positions simultaneously during `EndBlock`:
+
+```go
+// EndBlock
+// k is the x/leverage keeper and ctx contains information on the current block
+if ctx.BlockHeight() >= k.GetLastInterestBlock() + k.GetBorrowInterestEpoch() {
```

Comment:
Should we instead do modulo the current height? i.e. `BorrowInterestEpoch % currBlock == 0`